cli/tools/repl.rs

```diff
@@ -113,32 +99,83 @@ impl Completer for EditorHelper {
 
         if let Ok(get_properties_response) = get_properties_response {
           if let Some(result) = get_properties_response.get("result") {
-            let candidates = result
+            let property_names = result
               .as_array()
               .unwrap()
               .iter()
-              .filter_map(|r| {
-                let name = r.get("name").unwrap().as_str().unwrap().to_string();
-
-                if name.starts_with("Symbol(") {
-                  return None;
-                }
-
-                if name.starts_with(&suffix[1..]) {
-                  return Some(name);
-                }
-
-                None
-              })
+              .map(|r| r.get("name").unwrap().as_str().unwrap().to_string())
               .collect();
 
-            return Ok((pos - (suffix.len() - 1), candidates));
+            return property_names;
           }
         }
       }
     }
 
-    Ok((pos, Vec::new()))
+    Vec::new()
+  }
+}
+
+fn is_word_boundary(c: char) -> bool {
+  if c == '.' {
+    false
+  } else {
+    char::is_ascii_whitespace(&c) || char::is_ascii_punctuation(&c)
+  }
+}
+
+fn get_expr_from_line_at_pos(line: &str, cursor_pos: usize) -> &str {
+  let start = line[..cursor_pos]
+    .rfind(is_word_boundary)
+    .map_or_else(|| 0, |i| i);
+  let end = line[cursor_pos..]
+    .rfind(is_word_boundary)
+    .map_or_else(|| cursor_pos, |i| cursor_pos + i);
+
+  let word = &line[start..end];
+  let word = word.strip_prefix(is_word_boundary).unwrap_or(word);
+  let word = word.strip_suffix(is_word_boundary).unwrap_or(word);
+
+  word
+}
+
+impl Completer for EditorHelper {
+  type Candidate = String;
+
+  fn complete(
+    &self,
+    line: &str,
+    pos: usize,
+    _ctx: &Context<'_>,
+  ) -> Result<(usize, Vec<String>), ReadlineError> {
+    let expr = get_expr_from_line_at_pos(line, pos);
+
+    // check if the expression is in the form `obj.prop`
+    if let Some(index) = expr.rfind('.') {
+      let sub_expr = &expr[..index];
+      let prop_name = &expr[index + 1..];
+      let candidates = self
+        .get_expression_property_names(sub_expr)
+        .into_iter()
+        .filter(|n| !n.starts_with("Symbol(") && n.starts_with(prop_name))
+        .collect();
+
+      Ok((pos - prop_name.len(), candidates))
+    } else {
+      // combine results of declarations and globalThis properties
+      let mut candidates = self
+        .get_expression_property_names("globalThis")
+        .into_iter()
+        .chain(self.get_global_lexical_scope_names())
+        .filter(|n| n.starts_with(expr))
+        .collect::<Vec<_>>();
+
+      // sort and remove duplicates
+      candidates.sort();
+      candidates.dedup(); // make sure to sort first
```

Comment:
I kind of dislike `dedup()` because it requires remembering that the data needs to be sorted (which I needed to do here anyway). Should I keep this or perhaps use a hash set above instead?

Reply:
I think it's fine, there are comments so it's all good